File: go/internal/harness/harness.go
// Package harness implements the benchmark-binary protocol shared by every Go
// benchmark program. rust/common/src/harness.rs implements the same protocol;
// METHODOLOGY.md ("Measurement protocol") is the normative description.
//
//	<bin> list
//	<bin> run      <workload> [--param k=v]... [--warmup-iters N] [--warmup-min-ms N]
//	                          [--iters N] [--min-ms N] [--max-iters N] [--max-ms N] [--batch-min-ms N]
//	<bin> validate <workload> [--param k=v]...
//
// Setup, Prepare, Check and Teardown are never timed. Only Run is timed, with
// the monotonic clock. The result is a single JSON document on stdout; phase
// markers on stderr let the orchestrator align its RSS sampler with the
// measured window.
package harness

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"

	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/common"
)

const schemaVersion = 1

// Mode selects how Run is timed.
type Mode int

const (
	// ModeIter: one Run call is one timed iteration (a whole workload pass).
	ModeIter Mode = iota
	// ModeOp: one Run call is a single short operation. Ops are timed
	// individually into a histogram and grouped into batches for samples.
	ModeOp
)

func (m Mode) String() string {
	if m == ModeOp {
		return "op"
	}
	return "iter"
}

// Instance is a prepared workload. Run performs the timed work and returns a
// digest of its result; every call must return the same digest, otherwise
// the run is rejected as non-deterministic.
type Instance interface{ Run() uint64 }

// Preparer resets state before each timed Run (e.g. re-copying unsorted data).
type Preparer interface{ Prepare() }

// Checker computes an expensive digest of the post-Run state. It is called
// after the first and the last Run and both values must agree.
type Checker interface{ Check() uint64 }

// Teardowner releases resources after measurement (used by memory workloads
// to observe whether memory is returned to the OS).
type Teardowner interface{ Teardown() }

// ExtraReporter adds workload-specific metrics to the result.
type ExtraReporter interface{ Extra() map[string]any }

// Work describes how much logical work one Run call performs.
type Work struct {
	Unit       string  // e.g. "calls", "bytes", "elements", "ops"
	PerRun     float64 // units of work per Run call
	InputBytes int64   // size of the input consumed per Run call, if meaningful
}

// Workload is a named benchmark exposed by a binary.
type Workload struct {
	Name  string
	Mode  Mode
	Setup func(p Params) (Instance, Work, error)
}

type options struct {
	warmupIters, warmupMinMs, iters, minMs, maxIters, maxMs, batchMinMs int
}

// clock reads only the monotonic clock (one vDSO call), matching Rust's
// Instant::now(). time.Now() would additionally read the wall clock.
type clock struct{ base time.Time }

func (c clock) now() int64 { return int64(time.Since(c.base)) }

// Main runs the protocol for a binary exposing the given workloads.
func Main(bin string, workloads []Workload) {
	if len(os.Args) < 2 {
		usage(bin, workloads)
		os.Exit(2)
	}
	switch cmd := os.Args[1]; cmd {
	case "list":
		names := make([]map[string]string, 0, len(workloads))
		for _, w := range workloads {
			names = append(names, map[string]string{"name": w.Name, "mode": w.Mode.String()})
		}
		emit(map[string]any{"bin": bin, "lang": "go", "workloads": names})
	case "run", "validate":
		if len(os.Args) < 3 {
			usage(bin, workloads)
			os.Exit(2)
		}
		name := os.Args[2]
		var wl *Workload
		for i := range workloads {
			if workloads[i].Name == name {
				wl = &workloads[i]
			}
		}
		if wl == nil {
			fail(fmt.Errorf("unknown workload %q for %s", name, bin))
		}
		params := Params{}
		o := options{}
		fs := flag.NewFlagSet(cmd, flag.ExitOnError)
		fs.Var(paramFlag(params), "param", "workload parameter key=value (repeatable)")
		fs.IntVar(&o.warmupIters, "warmup-iters", 2, "minimum warmup iterations (batches in op mode)")
		fs.IntVar(&o.warmupMinMs, "warmup-min-ms", 500, "minimum warmup time")
		fs.IntVar(&o.iters, "iters", 5, "minimum measured iterations (batches in op mode)")
		fs.IntVar(&o.minMs, "min-ms", 1000, "minimum measured time")
		fs.IntVar(&o.maxIters, "max-iters", 100000, "hard cap on iterations per phase")
		fs.IntVar(&o.maxMs, "max-ms", 600000, "hard cap on wall time per phase")
		fs.IntVar(&o.batchMinMs, "batch-min-ms", 10, "op mode: target duration of one batch")
		_ = fs.Parse(os.Args[3:])
		if cmd == "validate" {
			o.warmupIters, o.warmupMinMs, o.iters, o.minMs = 0, 0, 1, 0
		}
		res, err := execute(bin, wl, params, o)
		if err != nil {
			fail(err)
		}
		emit(res)
	default:
		usage(bin, workloads)
		os.Exit(2)
	}
}

func usage(bin string, workloads []Workload) {
	fmt.Fprintf(os.Stderr, "usage: %s list | run <workload> [flags] | validate <workload> [flags]\nworkloads:", bin)
	for _, w := range workloads {
		fmt.Fprintf(os.Stderr, " %s", w.Name)
	}
	fmt.Fprintln(os.Stderr)
}

func emit(v any) {
	b, err := json.Marshal(v)
	if err != nil {
		fail(err)
	}
	os.Stdout.Write(append(b, '\n'))
}

func fail(err error) {
	b, _ := json.Marshal(map[string]any{"schema": schemaVersion, "lang": "go", "error": err.Error()})
	os.Stdout.Write(append(b, '\n'))
	os.Exit(1)
}

func phase(name string) { fmt.Fprintf(os.Stderr, "@@phase %s\n", name) }

type digestState struct {
	have   bool
	digest uint64
}

func (d *digestState) observe(x uint64) error {
	if !d.have {
		d.have, d.digest = true, x
		return nil
	}
	if x != d.digest {
		return fmt.Errorf("non-deterministic result digest: %016x != %016x", x, d.digest)
	}
	return nil
}

func execute(bin string, wl *Workload, params Params, o options) (res map[string]any, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("panic: %v", r)
		}
	}()
	inst, work, err := wl.Setup(params)
	if err != nil {
		return nil, err
	}
	rssSetup, hwmSetup := memStatus()

	var (
		clk              = clock{base: time.Now()}
		prep, _          = inst.(Preparer)
		checker, _       = inst.(Checker)
		ds               digestState
		checkFirst       uint64
		haveCheck        bool
		warm, samples    []int64
		opsPerSample     int64 = 1
		hist             *Histogram
		maxPhase         = int64(o.maxMs) * int64(time.Millisecond)
		warmupMin, minNs = int64(o.warmupMinMs) * 1e6, int64(o.minMs) * 1e6
	)
	afterFirst := func() {
		if checker != nil && !haveCheck {
			checkFirst, haveCheck = checker.Check(), true
		}
	}

	var snap0, snap1 rtSnapshot
	switch wl.Mode {
	case ModeIter:
		one := func() (int64, error) {
			if prep != nil {
				prep.Prepare()
			}
			t0 := clk.now()
			d := inst.Run()
			dt := clk.now() - t0
			if err := ds.observe(d); err != nil {
				return 0, err
			}
			afterFirst()
			return dt, nil
		}
		var sum int64
		start := clk.now()
		for i := 0; (i < o.warmupIters || sum < warmupMin) && i < o.maxIters && clk.now()-start < maxPhase; i++ {
			dt, err := one()
			if err != nil {
				return nil, err
			}
			warm = append(warm, dt)
			sum += dt
		}
		snap0 = snapshotRuntime()
		phase("measure_start")
		sum, start = 0, clk.now()
		for i := 0; (i < o.iters || sum < minNs) && i < o.maxIters && clk.now()-start < maxPhase; i++ {
			dt, err := one()
			if err != nil {
				return nil, err
			}
			samples = append(samples, dt)
			sum += dt
		}
		phase("measure_end")
		snap1 = snapshotRuntime()

	case ModeOp:
		batch := func(n int64, h *Histogram) (int64, error) {
			var s int64
			for k := int64(0); k < n; k++ {
				t0 := clk.now()
				d := inst.Run()
				dt := clk.now() - t0
				if err := ds.observe(d); err != nil {
					return 0, err
				}
				afterFirst()
				s += dt
				if h != nil {
					h.Record(uint64(dt))
				}
			}
			return s, nil
		}
		// Calibrate the batch size so one batch lasts about batchMinMs.
		var calNs, calOps int64
		calStart := clk.now()
		for calOps == 0 || clk.now()-calStart < 20*int64(time.Millisecond) {
			dt, err := batch(1, nil)
			if err != nil {
				return nil, err
			}
			calNs += dt
			calOps++
		}
		perOp := max(calNs/calOps, 1)
		opsPerSample = max(int64(o.batchMinMs)*1e6/perOp, 1)
		if o.iters == 1 && o.minMs == 0 { // validate mode: a single op suffices
			opsPerSample = 1
		}
		var sum int64
		start := clk.now()
		for i := 0; (i < o.warmupIters || sum < warmupMin) && i < o.maxIters && clk.now()-start < maxPhase; i++ {
			dt, err := batch(opsPerSample, nil)
			if err != nil {
				return nil, err
			}
			warm = append(warm, dt)
			sum += dt
		}
		hist = &Histogram{}
		snap0 = snapshotRuntime()
		phase("measure_start")
		sum, start = 0, clk.now()
		for i := 0; (i < o.iters || sum < minNs) && i < o.maxIters && clk.now()-start < maxPhase; i++ {
			dt, err := batch(opsPerSample, hist)
			if err != nil {
				return nil, err
			}
			samples = append(samples, dt)
			sum += dt
		}
		phase("measure_end")
		snap1 = snapshotRuntime()
	}

	checksum := common.Hex(ds.digest)
	if checker != nil {
		last := checker.Check()
		if haveCheck && last != checkFirst {
			return nil, fmt.Errorf("post-run check digest changed: %016x != %016x", last, checkFirst)
		}
		checksum += ":" + common.Hex(last)
	}
	rssEnd, hwmEnd := memStatus()

	var extra map[string]any
	if er, ok := inst.(ExtraReporter); ok {
		extra = er.Extra()
	}
	rss := map[string]int64{"after_setup": rssSetup, "hwm_after_setup": hwmSetup, "end": rssEnd, "hwm_end": hwmEnd}
	if td, ok := inst.(Teardowner); ok {
		td.Teardown()
		rss["after_teardown"], _ = memStatus()
	}

	res = map[string]any{
		"schema":         schemaVersion,
		"lang":           "go",
		"bin":            bin,
		"workload":       wl.Name,
		"mode":           wl.Mode.String(),
		"params":         params,
		"work":           map[string]any{"unit": work.Unit, "per_run": work.PerRun, "input_bytes": work.InputBytes},
		"checksum":       checksum,
		"warmup_ns":      nonNil(warm),
		"samples_ns":     nonNil(samples),
		"ops_per_sample": opsPerSample,
		"rss_kb":         rss,
		"rusage_measure": snap0.ru.delta(snap1.ru),
		"runtime":        runtimeDelta(snap0, snap1),
		"extra":          extra,
		"build":          buildInfo(),
	}
	if hist != nil {
		res["histogram"] = hist.export()
	}
	return res, nil
}

func nonNil(x []int64) []int64 {
	if x == nil {
		return []int64{}
	}
	return x
}

File: rust/common/src/harness.rs
//! The benchmark-binary protocol, identical to go/internal/harness/harness.go.
//!
//! ```text
//! <bin> list
//! <bin> run      <workload> [--param k=v]... [--warmup-iters N] [--warmup-min-ms N]
//!                           [--iters N] [--min-ms N] [--max-iters N] [--max-ms N] [--batch-min-ms N]
//! <bin> validate <workload> [--param k=v]...
//! ```
//!
//! Setup, prepare, check and teardown are never timed. Only `run` is timed,
//! with the monotonic clock. One JSON document is printed on stdout; phase
//! markers on stderr let the orchestrator align its RSS sampler with the
//! measured window. METHODOLOGY.md ("Measurement protocol") is normative.

use std::panic::{AssertUnwindSafe, catch_unwind};
use std::process;
use std::time::Instant;

use serde_json::{Map, Value, json};

use crate::alloc;
use crate::digest::hex;
use crate::hist::Histogram;
use crate::params::Params;
use crate::procfs;

pub const SCHEMA_VERSION: u32 = 1;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Mode {
    /// One `run` call is one timed iteration (a whole workload pass).
    Iter,
    /// One `run` call is one short operation, timed individually into a
    /// histogram and grouped into batches for samples.
    Op,
}

impl Mode {
    fn as_str(self) -> &'static str {
        match self {
            Mode::Iter => "iter",
            Mode::Op => "op",
        }
    }
}

/// A prepared workload. `run` performs the timed work and returns a digest of
/// its result; every call must return the same digest.
pub trait Instance {
    fn run(&mut self) -> u64;
    /// Untimed reset before each iteration (e.g. re-copying unsorted data).
    fn prepare(&mut self) {}
    /// Whether `check` is implemented.
    fn has_check(&self) -> bool {
        false
    }
    /// Untimed, expensive digest of the post-run state; called after the
    /// first and last run and both values must agree.
    fn check(&mut self) -> u64 {
        0
    }
    /// Untimed release of resources after measurement.
    fn teardown(&mut self) {}
    /// Whether `teardown` is implemented (the harness then samples RSS again).
    fn has_teardown(&self) -> bool {
        false
    }
    /// Workload-specific metrics.
    fn extra(&mut self) -> Option<Value> {
        None
    }
}

/// How much logical work one `run` call performs.
pub struct Work {
    pub unit: &'static str,
    pub per_run: f64,
    pub input_bytes: u64,
}

pub type SetupFn = fn(&Params) -> Result<(Box<dyn Instance>, Work), String>;

pub struct Workload {
    pub name: &'static str,
    pub mode: Mode,
    pub setup: SetupFn,
}

struct Options {
    warmup_iters: u64,
    warmup_min_ms: u64,
    iters: u64,
    min_ms: u64,
    max_iters: u64,
    max_ms: u64,
    batch_min_ms: u64,
}

impl Default for Options {
    fn default() -> Self {
        Options {
            warmup_iters: 2,
            warmup_min_ms: 500,
            iters: 5,
            min_ms: 1000,
            max_iters: 100_000,
            max_ms: 600_000,
            batch_min_ms: 10,
        }
    }
}

/// Run the protocol for a binary exposing `workloads`. Never returns.
pub fn main(bin: &str, workloads: &[Workload]) -> ! {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        usage(bin, workloads);
    }
    match args[1].as_str() {
        "list" => {
            let names: Vec<Value> =
                workloads.iter().map(|w| json!({"name": w.name, "mode": w.mode.as_str()})).collect();
            emit(&json!({"bin": bin, "lang": "rust", "workloads": names}));
            process::exit(0);
        }
        cmd @ ("run" | "validate") => {
            if args.len() < 3 {
                usage(bin, workloads);
            }
            let name = &args[2];
            let Some(wl) = workloads.iter().find(|w| w.name == name) else {
                fail(&format!("unknown workload {name:?} for {bin}"));
            };
            let (params, mut o) = parse_flags(&args[3..]);
            if cmd == "validate" {
                o.warmup_iters = 0;
                o.warmup_min_ms = 0;
                o.iters = 1;
                o.min_ms = 0;
            }
            let outcome = catch_unwind(AssertUnwindSafe(|| execute(bin, wl, &params, &o)));
            match outcome {
                Ok(Ok(v)) => {
                    emit(&v);
                    process::exit(0);
                }
                Ok(Err(e)) => fail(&e),
                Err(p) => {
                    let msg = p
                        .downcast_ref::<String>()
                        .cloned()
                        .or_else(|| p.downcast_ref::<&str>().map(|s| s.to_string()))
                        .unwrap_or_else(|| "unknown panic".to_string());
                    fail(&format!("panic: {msg}"));
                }
            }
        }
        _ => usage(bin, workloads),
    }
}

fn usage(bin: &str, workloads: &[Workload]) -> ! {
    let names: Vec<&str> = workloads.iter().map(|w| w.name).collect();
    eprintln!(
        "usage: {bin} list | run <workload> [flags] | validate <workload> [flags]\nworkloads: {}",
        names.join(" ")
    );
    process::exit(2);
}

fn parse_flags(args: &[String]) -> (Params, Options) {
    let mut params = Params::default();
    let mut o = Options::default();
    let mut i = 0;
    while i < args.len() {
        let (flag, inline) = match args[i].split_once('=') {
            Some((f, v)) if f.starts_with("--") => (f.to_string(), Some(v.to_string())),
            _ => (args[i].clone(), None),
        };
        let mut value = || -> String {
            if let Some(v) = &inline {
                return v.clone();
            }
            i += 1;
            args.get(i).cloned().unwrap_or_else(|| fail(&format!("flag {flag} needs a value")))
        };
        let num = |s: String| -> u64 {
            s.parse().unwrap_or_else(|_| fail(&format!("flag {flag}: bad number {s:?}")))
        };
        match flag.trim_start_matches('-') {
            "param" => {
                let kv = value();
                let Some((k, v)) = kv.split_once('=') else {
                    fail(&format!("want key=value, got {kv:?}"));
                };
                if k.is_empty() {
                    fail(&format!("want key=value, got {kv:?}"));
                }
                params.0.insert(k.to_string(), v.to_string());
            }
            "warmup-iters" => o.warmup_iters = num(value()),
            "warmup-min-ms" => o.warmup_min_ms = num(value()),
            "iters" => o.iters = num(value()),
            "min-ms" => o.min_ms = num(value()),
            "max-iters" => o.max_iters = num(value()),
            "max-ms" => o.max_ms = num(value()),
            "batch-min-ms" => o.batch_min_ms = num(value()),
            other => fail(&format!("unknown flag {other:?}")),
        }
        i += 1;
    }
    (params, o)
}

fn emit(v: &Value) {
    println!("{}", serde_json::to_string(v).expect("serialize result"));
}

fn fail(msg: &str) -> ! {
    emit(&json!({"schema": SCHEMA_VERSION, "lang": "rust", "error": msg}));
    process::exit(1);
}

fn phase(name: &str) {
    eprintln!("@@phase {name}");
}

#[derive(Default)]
struct DigestState {
    have: bool,
    digest: u64,
}

impl DigestState {
    #[inline]
    fn observe(&mut self, x: u64) -> Result<(), String> {
        if !self.have {
            self.have = true;
            self.digest = x;
            return Ok(());
        }
        if x != self.digest {
            return Err(format!("non-deterministic result digest: {x:016x} != {:016x}", self.digest));
        }
        Ok(())
    }
}

struct Snap {
    ru: procfs::Rusage,
    al: alloc::Snapshot,
}

fn snapshot() -> Snap {
    Snap { ru: procfs::rusage(), al: alloc::snapshot() }
}

fn build_info() -> Value {
    json!({
        "rustc_version": env!("BENCH_RUSTC_VERSION"),
        "profile": env!("BENCH_BUILD_PROFILE"),
        "opt_level": env!("BENCH_BUILD_OPT_LEVEL"),
        "debug": env!("BENCH_BUILD_DEBUG"),
        "target": env!("BENCH_BUILD_TARGET"),
        "target_features": env!("BENCH_TARGET_FEATURES"),
        "allocator": alloc::allocator_name(),
        "alloc_stats": alloc::counting_enabled(),
    })
}

fn execute(bin: &str, wl: &Workload, params: &Params, o: &Options) -> Result<Value, String> {
    let (mut inst, work) = (wl.setup)(params)?;
    let (rss_setup, hwm_setup) = procfs::mem_status();

    let has_check = inst.has_check();
    let mut ds = DigestState::default();
    let mut check_first: Option<u64> = None;
    let mut warm: Vec<u64> = Vec::new();
    // Pre-sized so harness bookkeeping does not allocate inside the measured window.
    let mut samples: Vec<u64> = Vec::with_capacity(4096);
    let mut ops_per_sample: u64 = 1;
    let mut hist: Option<Histogram> = None;
    let max_phase = o.max_ms as u128 * 1_000_000;
    let warmup_min = o.warmup_min_ms * 1_000_000;
    let min_ns = o.min_ms * 1_000_000;

    let snap0;
    let snap1;
    match wl.mode {
        Mode::Iter => {
            let mut one = |inst: &mut Box<dyn Instance>| -> Result<u64, String> {
                inst.prepare();
                let t0 = Instant::now();
                let d = inst.run();
                let dt = t0.elapsed().as_nanos() as u64;
                ds.observe(d)?;
                if has_check && check_first.is_none() {
                    check_first = Some(inst.check());
                }
                Ok(dt)
            };
            let mut sum = 0u64;
            let start = Instant::now();
            let mut i = 0u64;
            while (i < o.warmup_iters || sum < warmup_min)
                && i < o.max_iters
                && start.elapsed().as_nanos() < max_phase
            {
                let dt = one(&mut inst)?;
                warm.push(dt);
                sum += dt;
                i += 1;
            }
            phase("measure_start");
            snap0 = snapshot();
            alloc::reset_peak();
            let (mut sum, start, mut i) = (0u64, Instant::now(), 0u64);
            while (i < o.iters || sum < min_ns) && i < o.max_iters && start.elapsed().as_nanos() < max_phase {
                let dt = one(&mut inst)?;
                samples.push(dt);
                sum += dt;
                i += 1;
            }
            snap1 = snapshot();
            phase("measure_end");
        }
        Mode::Op => {
            let mut batch =
                |inst: &mut Box<dyn Instance>, n: u64, h: Option<&mut Histogram>| -> Result<u64, String> {
                    let mut s = 0u64;
                    let mut h = h;
                    for _ in 0..n {
                        let t0 = Instant::now();
                        let d = inst.run();
                        let dt = t0.elapsed().as_nanos() as u64;
                        ds.observe(d)?;
                        if has_check && check_first.is_none() {
                            check_first = Some(inst.check());
                        }
                        s += dt;
                        if let Some(h) = h.as_deref_mut() {
                            h.record(dt);
                        }
                    }
                    Ok(s)
                };
            // Calibrate the batch size so one batch lasts about batch_min_ms.
            let (mut cal_ns, mut cal_ops) = (0u64, 0u64);
            let cal_start = Instant::now();
            while cal_ops == 0 || cal_start.elapsed().as_nanos() < 20_000_000 {
                cal_ns += batch(&mut inst, 1, None)?;
                cal_ops += 1;
            }
            let per_op = (cal_ns / cal_ops).max(1);
            ops_per_sample = (o.batch_min_ms * 1_000_000 / per_op).max(1);
            if o.iters == 1 && o.min_ms == 0 {
                ops_per_sample = 1; // validate mode: a single op suffices
            }
            let (mut sum, start, mut i) = (0u64, Instant::now(), 0u64);
            while (i < o.warmup_iters || sum < warmup_min)
                && i < o.max_iters
                && start.elapsed().as_nanos() < max_phase
            {
                let dt = batch(&mut inst, ops_per_sample, None)?;
                warm.push(dt);
                sum += dt;
                i += 1;
            }
            let mut h = Histogram::new();
            phase("measure_start");
            snap0 = snapshot();
            alloc::reset_peak();
            let (mut sum, start, mut i) = (0u64, Instant::now(), 0u64);
            while (i < o.iters || sum < min_ns) && i < o.max_iters && start.elapsed().as_nanos() < max_phase {
                let dt = batch(&mut inst, ops_per_sample, Some(&mut h))?;
                samples.push(dt);
                sum += dt;
                i += 1;
            }
            snap1 = snapshot();
            phase("measure_end");
            hist = Some(h);
        }
    }

    let mut checksum = hex(ds.digest);
    if has_check {
        let last = inst.check();
        if let Some(first) = check_first
            && first != last
        {
            return Err(format!("post-run check digest changed: {last:016x} != {first:016x}"));
        }
        checksum = format!("{checksum}:{}", hex(last));
    }
    let (rss_end, hwm_end) = procfs::mem_status();
    let extra = inst.extra().unwrap_or(Value::Null);
    let mut rss = Map::new();
    rss.insert("after_setup".into(), json!(rss_setup));
    rss.insert("hwm_after_setup".into(), json!(hwm_setup));
    rss.insert("end".into(), json!(rss_end));
    rss.insert("hwm_end".into(), json!(hwm_end));
    if inst.has_teardown() {
        inst.teardown();
        rss.insert("after_teardown".into(), json!(procfs::mem_status().0));
    }
    let mut runtime = alloc::delta_json(&snap0.al, &snap1.al);
    if let Value::Object(m) = &mut runtime {
        m.insert("threads_end".into(), json!(procfs::thread_count()));
    }

    let mut res = json!({
        "schema": SCHEMA_VERSION,
        "lang": "rust",
        "bin": bin,
        "workload": wl.name,
        "mode": wl.mode.as_str(),
        "params": params.0,
        "work": {"unit": work.unit, "per_run": work.per_run, "input_bytes": work.input_bytes},
        "checksum": checksum,
        "warmup_ns": warm,
        "samples_ns": samples,
        "ops_per_sample": ops_per_sample,
        "rss_kb": Value::Object(rss),
        "rusage_measure": snap0.ru.delta_json(&snap1.ru),
        "runtime": runtime,
        "extra": extra,
        "build": build_info(),
    });
    if let Some(h) = hist {
        res["histogram"] = h.export();
    }
    Ok(res)
}
